src/core/generator/from_json.rs

```diff
@@ -0,0 +1,359 @@
+        self.generate_schema();
+    }
+}
+
+pub fn from_json(url: &str, json_resp: &Value) -> ConfigModule {
+    let mut ctx = ConfigGenerator::new(url);
+    ctx.generate(json_resp);
+    ConfigModule::from(ctx.config)
+}
+
+#[cfg(test)]
+mod test {
+    use serde_json::json;
+    use url::Url;
+
+    use crate::core::config::ConfigModule;
+    use crate::core::generator::from_json::{from_json, ConfigGenerator, UrlQueryParser};
+
+    #[test]
+    fn test_should_generate_type() {
+        let config_gen = ConfigGenerator::new("https://www.jsonplaceholder.typicode.come/users");
+        assert!(config_gen.should_generate_type(&json!("Testing")));
+        assert!(config_gen.should_generate_type(&json!(12)));
+        assert!(config_gen.should_generate_type(&json!(12.3)));
+        assert!(config_gen.should_generate_type(&json!(-12)));
+        assert!(config_gen.should_generate_type(&json!(-12.2)));
+        assert!(config_gen.should_generate_type(&json!(true)));
+        assert!(config_gen.should_generate_type(&json!(false)));
+        assert!(config_gen.should_generate_type(&json!([1, 2, 3])));
+        assert!(config_gen.should_generate_type(&json!({"name":"test", "age": 12})));
+
+        // ignore the empty types.
+        assert!(!config_gen.should_generate_type(&json!([])));
+        assert!(!config_gen.should_generate_type(&json!({})));
+
+        // not valid field names.
+        assert!(!config_gen.should_generate_type(&json!({"10": {
+            "name": "test",
+            "age": 12
+        }})));
+
+        assert!(!config_gen.should_generate_type(&json!({"user info": {
+            "age": 12
+        }})));
+    }
+
+    #[test]
+    fn test_generate_upstream() {
+        let mut config_gen =
+            ConfigGenerator::new("https://www.jsonplaceholder.typicode.come/users");
+        config_gen.generate_upstream();
+        assert!(config_gen.config.upstream.base_url.is_some())
+    }
+
+    #[test]
+    fn test_generation_of_config_with_incompatible_json_properties() {
+        let resp = r#"
+        {
+            "colors": [],
+            "campaignTemplates": {
+                "10": {
+                    "name": "test"
+                },
+                "15": {
+                    "name": "test"
+                }
+            }
+        }
+        "#;
+
+        let resp = serde_json::from_str(resp).unwrap();
+        let cgf_module = ConfigModule::from(from_json("https://example.com", &resp).config);
+        insta::assert_snapshot!(cgf_module.to_sdl());
+    }
+
+    #[test]
+    fn test_nested_same_json_properties() {
+        let resp = r#"
+            {
+                "container": {
+                    "name": "Testing",
+                    "container": {
+                        "name": "Testing",
+                        "container": {
+                            "age": 16
+                        }
+                    }
+                }
+            }
+        "#;
```

Comment:
- Writing JSON in Rust is a little tough, consider writing the inputs in `jsonc` or `YML` files for eg:

    ```json
    {
    "request": {
        "url": "http://example.com/api/endpoint"
    },
    "response": {
        "status": 200,
        "body": {}
    }
    }
    ```
- Use `datatest` to generate all the tests.

Reply:
done, have moved hardcoded json files to data tests.